### DISCUSSION TITLE
Appium Inspector - getPageSource NotImplementedError

### DISCUSSION THREAD

**Discussion by selen-oruc-alg** [appium/appium#18928] [Q&A] [answered]
Hello,

I need to inspect the Flutter application elements visually with Appium Inspector. When I try to create an Appium session with **Appium Inspector**, I get a `getPageSource` error. This issue only happens when I choose the automation name Flutter. I can connect the Appium session successfully via the Python script(which has the same desired caps and environment).

Could you please help me how to fix the issue or find a workaround? Thanks in advance :)

Here are my environment details:
[Appium](https://github.com/appium/appium): 2.0.0
[Appium Inspector](https://github.com/appium/appium-inspector): 2023.7.1
[Appium Flutter Driver](https://github.com/appium-userland/appium-flutter-driver): 1.19.1

Desired Capabilities:
```javascript
{
  "platformName": "Android",
  "appium:deviceName": "emulator-5554",
  "appium:automationName": "Flutter",
  "appium:app": "/<main-path>/flutter-debug.apk",
  "appium:retry_backoff_time": 2000,
  "appium:maxRetryCount": 5,
  "appium:autoGrantPermissions": "true"
}
``` 

Appium Server Logs:
```javascript
[HTTP] --> GET /wd/hub/session/<session>-<id>/source
[HTTP] {}
[debug] [FlutterDriver@9085 (<session>)] Calling AppiumDriver.getPageSource() with args: ["<session>-<id>"]
[debug] [FlutterDriver] Executing Flutter driver command 'getPageSource'
[debug] [FlutterDriver@9085 (<session>)] Encountered internal error running command: NotImplementedError: Not implemented yet for find.
[debug] [FlutterDriver@9085 (<session>)]     at FlutterDriver.getPageSource (/<main-path>/node_modules/@appium/base-driver/lib/basedriver/commands/find.js:67:13)
[debug] [FlutterDriver@9085 (<session>)]     at commandExecutor (/<main-path>/node_modules/@appium/base-driver/lib/basedriver/driver.js:122:18)
[debug] [FlutterDriver@9085 (<session>)]     at /<main-path>/node_modules/async-lock/lib/index.js:171:12
[debug] [FlutterDriver@9085 (<session>)]     at AsyncLock._promiseTry (/<main-path>/node_modules/async-lock/lib/index.js:304:31)
[debug] [FlutterDriver@9085 (<session>)]     at exec (/<main-path>/node_modules/async-lock/lib/index.js:170:9)
[debug] [FlutterDriver@9085 (<session>)]     at AsyncLock.acquire (/<main-path>/node_modules/async-lock/lib/index.js:187:3)
[debug] [FlutterDriver@9085 (<session>)]     at FlutterDriver.executeCommand (/<main-path>/node_modules/@appium/base-driver/lib/basedriver/driver.js:138:39)
[debug] [FlutterDriver@9085 (<session>)]     at processTicksAndRejections (node:internal/process/task_queues:95:5)
[debug] [FlutterDriver@9085 (<session>)]     at FlutterDriver.executeCommand (/<main-path>/.appium/node_modules/appium-flutter-driver/lib/driver.ts:184:16)
[debug] [FlutterDriver@9085 (<session>)]     at defaultBehavior (/opt/homebrew/lib/node_modules/appium/lib/appium.js:684:14)
[debug] [FlutterDriver@9085 (<session>)]     at AppiumDriver.executeWrappedCommand (/opt/homebrew/lib/node_modules/appium/lib/appium.js:774:16)
[debug] [FlutterDriver@9085 (<session>)]     at AppiumDriver.executeCommand (/opt/homebrew/lib/node_modules/appium/lib/appium.js:696:17)
[debug] [FlutterDriver@9085 (<session>)]     at asyncHandler (/opt/homebrew/lib/node_modules/appium/node_modules/@appium/base-driver/lib/protocol/protocol.js:393:19)
[HTTP] <-- GET /wd/hub/session/<session>-<id>/source 405 2 ms - 1439
``` 

Actually, the `getPageSource` method is implemented the throw an error that causes the error above([source](https://github.com/appium/appium/blob/master/packages/base-driver/lib/basedriver/commands/find.ts#L102)).
<img width="613" alt="image" src="https://github.com/appium/appium/assets/121935482/5ddc0b30-a5f7-4bf9-9238-2d224c99a5d5">

**Comment by KazuCocoa** [comment]
https://github.com/appium-userland/appium-flutter-driver#note does not have it in the flutter context as the implementation reason

**Comment by jlipps** [comment]
Unfortunately the Inspector is not designed to work with the Flutter driver.

**Comment by selen-oruc-alg** [comment]
Thanks a lot for your support @KazuCocoa @jlipps. 

Do you have a suggestion for an application that I can use instead of Appium Inspector? I tried [Flutter Inspector](https://docs.flutter.dev/tools/devtools/inspector) but it requires source code to inspect elements. I need to inspect elements with apk/app file.

**Reply by KazuCocoa** [reply, marked as answer]
Did you try `getRenderTree` addressed in https://github.com/appium-userland/appium-flutter-driver#note to get some hints?
It returns the response by the Flutter framework as-is. Elements listed in the tree can be used for flutter's finder

**Reply by rabiulsarker** [reply]
how can i use getRenderTree ? in the provided repo, i didn't find any workaround. @KazuCocoa  @jlipps

**Reply by jlipps** [reply]
there are examples in the readme. you may want to search the topic of "execute methods" in the main appium docs if you don't understand the examples.
